zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -922,108 +949,132 @@ where
         Some(sapling_dfvk.to_ovk(Scope::Internal))
     };
 
+    #[cfg(feature = "transparent-inputs")]
+    type TransparentMetadataT = TransparentAddressMetadata;
+    #[cfg(not(feature = "transparent-inputs"))]
+    type TransparentMetadataT = Infallible;
+
     #[cfg(feature = "orchard")]
-    let mut orchard_output_meta = vec![];
-    let mut sapling_output_meta = vec![];
-    let mut transparent_output_meta = vec![];
-    for (payment, output_pool) in proposal_step
-        .payment_pools()
-        .iter()
-        .map(|(idx, output_pool)| {
-            let payment = proposal_step
-                .transaction_request()
-                .payments()
-                .get(idx)
-                .expect(
-                    "The mapping between payment index and payment is checked in step construction",
-                );
-            (payment, output_pool)
-        })
-    {
-        let recipient_address: Address = payment
-            .recipient_address()
+    let mut orchard_output_meta: Vec<(
+        Recipient<_, PoolType, _>,
+        NonNegativeAmount,
+        Option<MemoBytes>,
+    )> = vec![];
+    let mut sapling_output_meta: Vec<(
+        Recipient<_, PoolType, _>,
+        NonNegativeAmount,
+        Option<MemoBytes>,
+    )> = vec![];
+    let mut transparent_output_meta: Vec<(
+        Recipient<_, _, ()>,
+        TransparentAddress,
+        Option<TransparentMetadataT>,
+        NonNegativeAmount,
+        StepOutputIndex,
+    )> = vec![];
+
+    for (&payment_index, output_pool) in proposal_step.payment_pools() {
+        let payment = proposal_step
+            .transaction_request()
+            .payments()
+            .get(&payment_index)
+            .expect(
+                "The mapping between payment index and payment is checked in step construction",
+            );
+        let recipient_address = payment.recipient_address();
+
+        let mut add_sapling_output = |builder: &mut Builder<_, _>,
+                                      to: sapling::PaymentAddress|
+         -> Result<(), ErrorT<DbT, InputsErrT, FeeRuleT>> {
+            let memo = payment.memo().map_or_else(MemoBytes::empty, |m| m.clone());
+            builder.add_sapling_output(sapling_external_ovk, to, payment.amount(), memo.clone())?;
+            sapling_output_meta.push((
+                Recipient::External(recipient_address.clone(), PoolType::SAPLING),
+                payment.amount(),
+                Some(memo),
+            ));
```

Comment:
Style nit: if you're going to mutate `sapling_output_meta` here without passing it in, why also pass in the `&mut Builder`? Either pass both, or close over both to mutate them.

Reply:
The reason for this is that the builder is used by more than one closure, which the borrow checker doesn't like if it is captured. However, since part of the purpose of these closures is to update the `*_meta` vectors, I agree they should be explicitly passed in.